crates/runtime/src/backtrace.rs

```diff
@@ -0,0 +1,260 @@
+                    return;
+                }
+                let fp = *(*state.limits).last_wasm_exit_fp.get();
+                (pc, fp)
+            }
+        };
+
+        // Trace through the first contiguous sequence of Wasm frames on the
+        // stack.
+        if let ControlFlow::Break(()) = Self::trace_through_wasm(
+            last_wasm_exit_pc,
+            last_wasm_exit_fp,
+            *(*state.limits).last_wasm_entry_sp.get(),
+            &mut f,
+        ) {
+            return;
+        }
+
+        // And then trace through each of the older contiguous sequences of Wasm
+        // frames on the stack.
+        for state in state.iter() {
+            // If the old state's PC is zero, then we have finished iterating
+            // over all sequences of contiguous Wasm frames.
+            if state.old_last_wasm_exit_pc == 0 {
+                debug_assert!(state.prev.get().is_null());
+                return;
+            }
+
+            if let ControlFlow::Break(()) = Self::trace_through_wasm(
+                state.old_last_wasm_exit_pc,
+                state.old_last_wasm_exit_fp,
+                state.old_last_wasm_entry_sp,
+                &mut f,
+            ) {
+                return;
+            }
+        }
+    }
+
+    /// Walk through a contiguous sequence of Wasm frames starting with the
+    /// frame at the given PC and FP and ending at `first_wasm_sp`.
+    unsafe fn trace_through_wasm(
+        mut pc: usize,
+        mut fp: usize,
+        first_wasm_sp: usize,
+        mut f: impl FnMut(Frame) -> ControlFlow<()>,
+    ) -> ControlFlow<()> {
+        log::trace!("=== Tracing through contiguous sequence of Wasm frames ===");
+        log::trace!("first_wasm_sp = 0x{:016x}", first_wasm_sp);
+        log::trace!("   initial pc = 0x{:016x}", pc);
+        log::trace!("   initial fp = 0x{:016x}", fp);
+
+        // In our host-to-Wasm trampoline, we save `-1` as a sentinal SP
+        // value for when the callee is not actually a core Wasm
+        // function (as determined by looking at the callee `vmctx`). If
+        // we encounter `-1`, this is an empty sequence of Wasm frames
+        // where a host called a host function so the following
+        // happened:
+        //
+        // * We entered the host-to-wasm-trampoline, saved (an invalid
+        //   sentinal for) entry SP, and tail called to the "Wasm"
+        //   callee,
+        //
+        // * entered the Wasm-to-host trampoline, saved the exit FP and
+        //   PC, and tail called to the host callee,
+        //
+        // * and are now in host code.
+        //
+        // Ultimately, this means that there are 0 Wasm frames in this
+        // contiguous sequence of Wasm frames, and we have nothing to
+        // walk through here.
+        if first_wasm_sp == -1_isize as usize {
+            log::trace!("Empty sequence of Wasm frames");
+            return ControlFlow::Continue(());
+        }
```

Comment:
I don't think so? Because the host has no guarantees of preserving frame pointers and the FP is the host's FP in this situation, so I don't think we can assert any sort of relation here.